crypto/asn1/tasn_dec.c

```diff
@@ -1119,25 +1135,22 @@ static int asn1_check_tlen(long *olen, int *otag, unsigned char *oclass,
              * If definite length, and no error, length + header can't exceed
              * total amount of data available.
              */
-            if (!(i & 0x81) && ((plen + ctx->hdrlen) > len)) {
+            if ((i & 0x81) == 0 && (plen + ctx->hdrlen) > len) {
                 ERR_raise(ERR_LIB_ASN1, ASN1_R_TOO_LONG);
                 asn1_tlc_clear(ctx);
                 return 0;
             }
         }
     }
 
-    if (i & 0x80) {
-        ERR_raise(ERR_LIB_ASN1, ASN1_R_BAD_OBJECT_HEADER);
-        asn1_tlc_clear(ctx);
-        return 0;
-    }
+    if ((i & 0x80) != 0)
+        goto err;
```

Comment:
Should there be just asn1_tlc_clear() in the err: block so you could reuse it by the other various return 0 calls here.

Reply:
Good thought - done.
This also fixes the issue below.